heapster-saw/src/Verifier/SAW/Heapster/Implication.hs

```diff
@@ -6698,194 +6666,268 @@ solveForArrayH ps _ mb_ap =
 
-      if bvEq off (llvmArrayOffset ap_lhs) && bvEq len (llvmArrayLen ap_lhs) then
-        proveVarLLVMArray_FromArray2 x ap_lhs len bs mb_ap >>>= \_ ->
-        return ()
-      else
-        implLLVMArrayGet x ap_lhs off len >>>= \ap_lhs' ->
-        recombinePerm x (ValPerm_LLVMArray ap_lhs') >>>
-        proveVarLLVMArray_FromArray2 x (llvmMakeSubArray ap_lhs off len) len bs mb_ap >>>= \_ ->
-        return ()
+    if bvEq off (llvmArrayOffset ap_lhs) && bvEq len (llvmArrayLen ap_lhs) then
+      proveVarLLVMArray_FromArray x ap_lhs len bs mb_ap
+    else
+      implLLVMArrayGet x ap_lhs off len >>>= \ap_lhs' ->
+      recombinePerm x (ValPerm_LLVMArray ap_lhs') >>>
+      proveVarLLVMArray_FromArray x (llvmMakeSubArray ap_lhs off len) len bs mb_ap
   where
-    containsRHS ::
+    -- Test if an atomic permission is a "suitable" array permission for the
+    -- given offset, length, and stride, meaning that it has the required
+    -- stride, could contain the offset and length, and does not have all of the
+    -- offset and length borrowed
+    suitableAP ::
       (1 <= w, KnownNat w) =>
-      PermExpr (BVType w) -> PermExpr (BVType w) -> AtomicPerm (LLVMPointerType w) -> Bool
-    containsRHS off len p =
-      case p of
-        Perm_LLVMArray ap ->
-          -- Return `True` if this permission *could* cover the desired off/len
-          all bvPropCouldHold $
-            bvPropRangeSubset (BVRange off len) $
-            BVRange (llvmArrayOffset ap) (llvmArrayLengthBytes ap)
-        _ -> False
-
--- Check if there's a block with array shape and eliminate it. This new array
--- could then be used in the case above, so recurse.
-proveVarLLVMArrayH x p psubst ps mb_ap
-  | Just i <- findIndex arrayBlock ps
-  , Perm_LLVMBlock bp <- ps !! i
-  , Just ap <- llvmBlockPermToArray bp =
-    implExtractSwapConjM x ps i >>>
-    implElimLLVMBlock x bp >>>
-    implSwapInsertConjM x (Perm_LLVMArray ap) (deleteNth i ps) i >>>
-    proveVarLLVMArrayH x p psubst (replaceNth i (Perm_LLVMArray ap) ps) mb_ap
-  where
-    arrayBlock (Perm_LLVMBlock bp)
-      | PExpr_ArrayShape {} <- llvmBlockShape bp = True
-    arrayBlock _ = False
-
-proveVarLLVMArrayH x p psubst ps mb_ap
-  | Just i <- findIndex seqFieldBlock ps =
+      PermExpr (BVType w) -> PermExpr (BVType w) -> Bytes ->
+      AtomicPerm (LLVMPointerType w) -> Bool
+    suitableAP off len stride (Perm_LLVMArray ap) =
+      -- Test that the strides are equal
+      llvmArrayStride ap == stride &&
+      -- Make sure the range [off,len) is not fully borrowed
+      not (llvmArrayRangeIsBorrowed ap (BVRange off len)) &&
+      -- Test if this permission *could* cover the desired off/len
+      all bvPropCouldHold (bvPropRangeSubset (BVRange off len)
+                                             (llvmArrayAbsOffsets ap))
+    suitableAP _ _ _ _ = False
+
+-- Check if there is a block that contains the required offset and length, in
+-- which case eliminate it
```

Comment:
Nope, this can be case 4 (eliminate a memblock to an array), or it can generate a bunch of array and/or field permissions that can be used for cases 1-3. I added a comment about how the four cases in the overall description do not correspond to the cases of the function. Lmk if it makes sense.